src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,182 @@
+    public CustomMessageHandlerService(ICustomMessageHandlerFactory customMessageHandlerFactory)
+    {
+        _customMessageHandlerFactory = customMessageHandlerFactory;
+    }
+
+    public async ValueTask<string> HandleCustomMessageAsync(
+        Solution solution,
+        string assemblyFolderPath,
+        string assemblyFileName,
+        string typeFullName,
+        string jsonMessage,
+        DocumentId? documentId,
+        CancellationToken cancellationToken)
+    {
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        _ = assemblyFolderPath ?? throw new ArgumentNullException(nameof(assemblyFolderPath));
+        _ = assemblyFileName ?? throw new ArgumentNullException(nameof(assemblyFileName));
+        _ = typeFullName ?? throw new ArgumentNullException(nameof(typeFullName));
+        _ = jsonMessage ?? throw new ArgumentNullException(nameof(jsonMessage));
+
+        var assemblyPath = Path.Combine(assemblyFolderPath, assemblyFileName);
+
+        Document? document = null;
+        if (documentId is not null)
+        {
+            document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(document);
+        }
+
+        try
+        {
+            ICustomMessageHandlerWrapper? handler;
+            lock (_lockObject)
+            {
+                // Check if the assembly is already loaded.
+                if (!_extensions.TryGetValue(assemblyFolderPath, out var extension))
+                {
+                    var loadContext = new System.Runtime.Loader.AssemblyLoadContext(name: $"RemoteCustomMessageHandlerService assembly load context for {assemblyFolderPath}", isCollectible: true);
+                    loadContext.Resolving += ResolveExtensionAssembly;
+
+                    extension = new CustomMessageHandlerExtension(loadContext);
+                    _extensions[assemblyFolderPath] = extension;
+                }
+
+                if (!extension.Handlers.TryGetValue(typeFullName, out handler))
+                {
+                    var assembly = extension.AssemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
+                    var type = assembly.GetType(typeFullName)
+                        ?? throw new InvalidOperationException($"Cannot find type {typeFullName} in {assemblyPath}.");
+
+                    handler = _customMessageHandlerFactory.Create(type);
+                }
+            }
+
+            var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+
+            var result = await handler.ExecuteAsync(message, document, solution, cancellationToken)
+                .ConfigureAwait(false);
+
+            var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+
+            return responseJson;
+        }
+        catch (Exception e)
+        {
+            Logger.Log(
```

Comment:
this will log telemetry IIRC on every plugin exception.  For a misbehaving plugin that makes requests and fails on every character typed, this could spam telemetry.

Would it be possible to split the exception handling for loading the plugin, and making the request?  Loading is more likely to be our issue, but failure in deserialization/handle/serialization is less likely to be something we need to record every instance of.

We already do have aggregate telemetry for request failures in LSP.  We will likely want to eventually update this for plugins too (currently all plugin failures would be bucketed under the handler name - https://github.com/dotnet/roslyn/blob/main/src/LanguageServer/Protocol/Handler/Telemetry/RequestTelemetryLogger.cs - but TBD based on final shape

Reply:
I limited logging to the first attempt to instantiate each handler.
Other exceptions are bubbled up to the extension. We can add more aggregate logging later if we want to.
Also we have built-in aggregate telemetry for API calls in Gladstone.